server/src/main/scala/services/graphstages/DynamicWorkloadCalculator.scala

```diff
@@ -0,0 +1,135 @@
+    val splitRatios: Set[ApiPaxTypeAndQueueCount] = splitsToUse.splitStyle match {
+      case UndefinedSplitStyle => splitsToUse.splits.map(qc => qc.copy(paxCount = 0))
+      case PaxNumbers =>
+        val splitsWithoutTransit = splitsToUse.splits.filter(_.queueType != Queues.Transfer)
+        val totalSplitsPax: Load = splitsWithoutTransit.toList.map(_.paxCount).sum
+        if (totalSplitsPax == 0.0)
+          splitsWithoutTransit
+        else
+          splitsWithoutTransit.map(qc => qc.copy(paxCount = qc.paxCount / totalSplitsPax))
+      case _ => splitsToUse.splits.map(qc => qc.copy(paxCount = qc.paxCount / 100))
+    }
+    splitRatios
+  }
+
+}
+
+
+case class DynamicWorkloadCalculator(defaultProcTimes: Map[Terminal, Map[PaxTypeAndQueue, Double]],
+                                     queueStatusProvider: QueueStatusProvider)
+  extends WorkloadCalculatorLike {
+
+  val log: Logger = LoggerFactory.getLogger(getClass)
+
+  override def flightLoadMinutes(flights: FlightsWithSplits): SplitMinutes = {
+    val minutes = new SplitMinutes
+
+    flightsWithPcpWorkload(combineCodeShares(flights.flights.values))
+      .foreach(incoming => minutes ++= flightToFlightSplitMinutes(incoming))
+
+    minutes
+  }
+
+  def flightToFlightSplitMinutes(flightWithSplits: ApiFlightWithSplits): Iterable[FlightSplitMinute] =
+    defaultProcTimes.get(flightWithSplits.apiFlight.Terminal) match {
+      case None => Iterable()
+      case Some(procTimes) =>
+        val flight = flightWithSplits.apiFlight
+
+        flightWithSplits.bestSplits.map(splitsToUse => {
+          val paxTypeAndQueueCounts = paxTypeAndQueueCountsFromSplits(splitsToUse)
+
+          val paxTypesAndQueuesMinusTransit = paxTypeAndQueueCounts.filterNot(_.queueType == Queues.Transfer)
+
+          flight.paxDeparturesByMinute(20)
+            .flatMap {
+              case (minuteMillis, flightPaxInMinute) =>
+                paxTypesAndQueuesMinusTransit
+                  .map { case ptqc@ApiPaxTypeAndQueueCount(pt, queue, _, _, _) =>
+                    def findAlternativeQueue(originalQueue: Queue, queuesToTry: Iterable[Queue]): Queue = {
+                      queuesToTry.find(queueStatusProvider.statusAt(flight.Terminal, _, SDate(minuteMillis).getHours()) == Open) match {
+                        case Some(queue) => queue
+                        case None =>
+                          log.error(s"Failed to find alternative queue for $pt. Reverting to $originalQueue")
+                          originalQueue
+                      }
+                    }
+
+                    val finalPtqc = queueStatusProvider.statusAt(flight.Terminal, queue, SDate(minuteMillis).getHours()) match {
+                      case Closed =>
+                        val newQueue = (queue, pt) match {
+                          case (EGate, EeaMachineReadable | EeaNonMachineReadable | EeaBelowEGateAge) =>
```

Comment:
I'm wondering if these fallback rules should be more clearly identified somehow, I'm not entirely sure how, but maybe at least extracting this function to something named "queueFallbackStrategy" or something similar.

Reply:
Yes good suggestion. Maybe a case class `QueueFallbacks(queue, fallbackQueues)` and then defined outside this file